### ISSUE TITLE
vine: Future tasks underperform while tasks are linearly dependent

### ISSUE THREAD

**Issue by JinZhou5042** [cooperative-computing-lab/cctools#3649]
Both FuturePythonTask and FutureFunctionCall offer big advantages to the users to build their applications. However, for the showcased performance problem for Future tasks in #3607, the Future tasks performance suffer exceptionally in case of large amount of tasks, with each task having its inputs depending on a prior task's outputs. 

Here is a script that can repeat the observation:

``` Python
#! /usr/bin/env python
import argparse
import ndcctools.taskvine as vine
import time

def my_func(num):
    return num + 1


def main():
    parser = argparse.ArgumentParser(description="")
    parser.add_argument('--tasks', type=int, default=10)
    parser.add_argument("--pythontask", action='store_true')
    parser.add_argument("--functioncall", action='store_true')
    parser.add_argument("--future", action='store_true')
    args = parser.parse_args()
    
    # Create Executor
    executor = vine.Executor(manager_name='test_manager', factory=False)
    executor.manager.declare_poncho("env.tar.gz")
    print("listening on port {}".format(executor.manager.port))

    # Create library task
    if args.functioncall:
        print("creating library from functions...")
        libtask = executor.create_library_from_functions('test-library', my_func, import_modules=None)
        # Install library on executor.manager
        executor.install_library(libtask)

    tasks = args.tasks

    data = 0

    time_start = time.time()
    if args.future:
        print(f"submitting {tasks} future tasks...")
    else:
        print(f"submitting {tasks} regular tasks...")


    for _ in range(tasks):
        if args.functioncall:
            if args.future:
                t = executor.future_funcall('test-library', 'my_func', data)
                data = executor.submit(t)
            else:
                t = vine.FunctionCall('test-library', 'my_func', data)
                executor.manager.submit(t)
                executor.manager.wait_for_task_id(t.id)
                data = t.output
        if args.pythontask:
            if args.future:
                t = executor.future_pythontask(my_func, data)
                data = executor.submit(t)
            else:
                t = vine.PythonTask(my_func, data)
                executor.manager.submit(t)
                executor.manager.wait_for_task_id(t.id)
                data = t.output
                pass

    if args.future:
        res = data.result()
    else:
        res = data
    print(f"all tasks are submitted")
    print(f"all tasks are done, res = {res}, time used = {time.time() - time_start}")


if __name__ == '__main__':
    main()


# vim: set sts=4 sw=4 ts=4 expandtab ft=python:

```

Without using `FutureFunctionCall`, the way to submit 20 tasks is to create one task, submit it, wait for its completion and retrieve the output as the next task's input, and repeat for 20 times. Here is the command for regular `FunctionCall` tasks:
``` python
python performance_future.py --tasks 20 --functioncall
```

By using `FutureFunctionCall` tasks, we can make a one-time submission, and the manager will hook up the output of one task and the input of another, thus we can make a one-time submission, and the manager will take care of scheduling all tasks by correctly surfacing their dependencies. By using this command can submit `FutureFunctionCall` tasks:
``` python
python performance_future.py --tasks 20 --functioncall --future
```

In case of tens of tasks, there is no remarkable difference between `FunctionCall` and `FutureFunctionCall` tasks. 

However, as I increased the number of tasks, for instance, 300 tasks, the one using `FunctionCall` consumed **5.8s**, the one with `FutureFunctionCall` took **133.4s**. The larger the number of tasks, the bigger the orders of magnitude differences.

**Comment by JinZhou5042** [comment]
As @BarrySlyDelgado has pointed out, I printed some auxiliary information in `vine_manager_check_inputs_available` which is invoked in `send_one_task`.

According to the printed information, for those `FunctionCall` tasks, since they were created and submitted one by one, one task was not able to be put into the task table until the previous one has done. The manager waits for the input file to be prepared for the first task, once it was done, the following tasks were executed smoothly.

![Screenshot from 2024-02-07 16-19-38](https://github.com/cooperative-computing-lab/cctools/assets/142265839/5751473c-b215-4f18-8fd2-9ff334d77169)

For `FutureFunctionCall` tasks, they were put into the task table together, and every time the manager wants to pick up a task to run, it will traverse all the available tasks in its table.

![Screenshot from 2024-02-07 16-23-27](https://github.com/cooperative-computing-lab/cctools/assets/142265839/bd4d1775-59af-4b72-a2c0-b4fd083ad50e)

The complexity of scheduling `FunctionCall` tasks is O(n) whereas O(n^2) for `FutureFunctionCall`s.

**Comment by dthain** [comment]
Yes, I see why we get this slowdown: all of blocked tasks are sitting in the scheduler, and get considered every time it is looking for a task to run.  In the big picture, the solution is to segregate the not-ready tasks from the ready tasks, so that the scheduler doesn't iterate over things that can't run anywhere.

This is in part why we separate the layer of "dag manager" (Makeflow, Parsl, Dask, etc) from the layer of "task scheduler" (Work Queue, TaskVine, etc) when designing a system stack.

**Comment by JinZhou5042** [comment]
No longer an issue
